File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/domain/models/ExampleModel.java
package com.oddly.ddd.domain.models;

import com.oddly.ddd.domain.models.infra.BaseModel;

/**
 * Business Model Object (BMO) for Example domain.
 * BMOs:
 * - Live in /domain/models/
 * - Have Model or BMO suffix
 * - Contain business logic and invariants
 * - MUST NOT have database/ORM attributes
 * - Pure domain logic only
 */
public class ExampleModel extends BaseModel {
    private String m_name;
    private String m_description;
    private String m_category;
    private boolean m_isActive;
    private String m_ownerId;

    public ExampleModel(String p_name, String p_description, String p_category, String p_ownerId) {
        super();
        this.m_name = p_name;
        this.m_description = p_description;
        this.m_category = p_category;
        this.m_isActive = true;
        this.m_ownerId = p_ownerId;
    }

    public ExampleModel(String p_id, String p_name, String p_description, 
                        String p_category, boolean p_isActive, String p_ownerId) {
        super(p_id);
        this.m_name = p_name;
        this.m_description = p_description;
        this.m_category = p_category;
        this.m_isActive = p_isActive;
        this.m_ownerId = p_ownerId;
    }

    // Business logic methods

    /**
     * Activates the example.
     * Business rule: Only the owner can activate.
     */
    public void activate(String p_userId) {
        validateOwnership(p_userId);
        this.m_isActive = true;
        touch();
    }

    /**
     * Deactivates the example.
     * Business rule: Only the owner can deactivate.
     */
    public void deactivate(String p_userId) {
        validateOwnership(p_userId);
        this.m_isActive = false;
        touch();
    }

    /**
     * Updates the example details.
     * Business rule: Only the owner can update.
     */
    public void update(String p_name, String p_description, String p_category, String p_userId) {
        validateOwnership(p_userId);
        
        if (p_name != null && !p_name.isBlank()) {
            this.m_name = p_name;
        }
        if (p_description != null) {
            this.m_description = p_description;
        }
        if (p_category != null) {
            this.m_category = p_category;
        }
        
        touch();
    }

    /**
     * Validates that the user owns this example.
     */
    private void validateOwnership(String p_userId) {
        if (!this.m_ownerId.equals(p_userId)) {
            throw new IllegalStateException("User does not own this example");
        }
    }

    @Override
    public void validate() {
        super.validate();
        
        if (m_name == null || m_name.isBlank()) {
            throw new IllegalStateException("Example name cannot be empty");
        }
        if (m_name.length() < 3 || m_name.length() > 100) {
            throw new IllegalStateException("Example name must be between 3 and 100 characters");
        }
        if (m_ownerId == null || m_ownerId.isBlank()) {
            throw new IllegalStateException("Example must have an owner");
        }
    }

    // Getters
    public String getName() {
        return m_name;
    }

    public String getDescription() {
        return m_description;
    }

    public String getCategory() {
        return m_category;
    }

    public boolean isActive() {
        return m_isActive;
    }

    public String getOwnerId() {
        return m_ownerId;
    }
}

File: infrastructures/ddd/typescript/src/domain/models/ExampleModel.ts
import { BaseModel } from './infra/BaseModel';
import { randomUUID } from 'crypto';

/**
 * Example Business Model Object (BMO) demonstrating domain layer patterns.
 * This class:
 * - Lives in /domain/models/
 * - Has Model or BMO suffix
 * - Contains business logic and invariants
 * - MUST NOT have database/ORM attributes
 * - Uses member field naming convention: _variable
 * - Uses parameter naming convention: p_variable
 */
export class ExampleModel extends BaseModel {
  private _name: string;
  private _description: string;
  private _ownerId: string;
  private _isActive: boolean;

  public get Name(): string {
    return this._name;
  }

  public get Description(): string {
    return this._description;
  }

  public get OwnerId(): string {
    return this._ownerId;
  }

  public get IsActive(): boolean {
    return this._isActive;
  }

  // Private constructor for hydration from database
  private constructor() {
    super();
    this._name = '';
    this._description = '';
    this._ownerId = '';
    this._isActive = false;
  }

  /**
   * Public factory method for creating new instances.
   * Validates business rules before returning.
   */
  public static Create(p_name: string, p_description: string, p_ownerId: string): ExampleModel {
    const model = new ExampleModel();
    model._id = randomUUID();
    model._name = p_name;
    model._description = p_description;
    model._ownerId = p_ownerId;
    model._isActive = true;
    model._createdAt = new Date();
    model._updatedAt = new Date();

    model.Validate();
    return model;
  }

  /**
   * Factory method for hydration from persistence layer.
   * Used by repositories when loading from database.
   */
  public static Hydrate(
    p_id: string,
    p_name: string,
    p_description: string,
    p_ownerId: string,
    p_isActive: boolean,
    p_createdAt: Date,
    p_updatedAt: Date
  ): ExampleModel {
    const model = new ExampleModel();
    model._id = p_id;
    model._name = p_name;
    model._description = p_description;
    model._ownerId = p_ownerId;
    model._isActive = p_isActive;
    model._createdAt = p_createdAt;
    model._updatedAt = p_updatedAt;

    return model;
  }

  // Business logic methods

  /**
   * Updates the details of the example.
   * Validates business rules before applying changes.
   */
  public UpdateDetails(p_name: string, p_description: string): void {
    if (!p_name || p_name.trim().length === 0) {
      throw new Error('Name cannot be empty');
    }

    this._name = p_name;
    this._description = p_description;
    this._updatedAt = new Date();

    this.Validate();
  }

  /**
   * Activates the example.
   */
  public Activate(): void {
    this._isActive = true;
    this._updatedAt = new Date();
  }

  /**
   * Deactivates the example.
   */
  public Deactivate(): void {
    this._isActive = false;
    this._updatedAt = new Date();
  }

  /**
   * Validates that the user owns this resource.
   * Used by ownership middleware and service layer.
   */
  public ValidateOwnership(p_userId: string): void {
    if (this._ownerId !== p_userId) {
      throw new Error('User does not own this resource');
    }
  }

  /**
   * Validates business invariants.
   * Override from BaseModel to add domain-specific rules.
   */
  public override Validate(): void {
    super.Validate();

    if (!this._name || this._name.trim().length === 0) {
      throw new Error('Example name cannot be empty');
    }

    if (this._name.length > 100) {
      throw new Error('Example name cannot exceed 100 characters');
    }

    if (!this._ownerId || this._ownerId.trim().length === 0) {
      throw new Error('Example must have an owner');
    }
  }
}
